boards: ti: sk_am62: Add OpenOCD support

dnltz opened the pull request:
This patch series adds an additional parameter to the openocd runner `--gdb-port-offset` and all required files to run `west debug` with the TI AM62 EVM.

dnltz committed:
boards: ti: sk_am62: Add OpenOCD support

The SK AM62 EVM is equipped with an XDS110 JTAG debugger. Add
missing OpenOCD configuration and some documentation to support
'west debug'.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz committed:
scripts: west: runner: openocd: Add gdb-client-port

The gdb-port defines the GDB port and the openocd runner passes
this value to openocd as well as gdb. However, the TI AM62x board
provides multiple ports for each subsystem. For example, systick
appears at 3333, A53 as 3334-3337, R5F as 3338 and M4F as 3339.

If we want to connect to the M4F, we need to add another port which
is different to the to the gdb-port value.

This patch adds an additional argument --gdb-client-port to define
the port which GDB should connect to. It defaults to 3333, identical
to gdb-port.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented:
@vaishnavachath I tested this on our kit which very identical to the EVM. Could you do me the favor and test this on the TI EVM? I currently have no access to our TI EVM and just wanna make sure it works ;)

vaishnavachath commented:
Sure @dnltz , I will test tomorrow on AM62 TI SK and let you know. Thank you for the contribution

gramsay0 commented on boards/ti/sk_am62/board.cmake:
```diff
@@ -0,0 +1,13 @@
+  board_set_debugger_ifnset(openocd)
+  board_set_flasher_ifnset(openocd)
+
+  board_runner_args(openocd "--no-init" "--no-halt" "--no-targets" "--gdb-port-offset=6")
```
Good work!
We have done something similar on A53 cores, although not integrated into zephyr/west.
What do you think about specifying the port rather than an offset? e.g:
`--gdb-port=3333 --gdb-client-port=3339`

dnltz replied:
yeah, that's actually not a bad idea. Thanks!

gramsay0 commented:
I haven't tried this yet, but I remember we had issues with software breakpoints and used `gdb_breakpoint_override hard`.
There were also issues with openocd versioning, I think it required >= 0.12

dnltz commented:
> I haven't tried this yet, but I remember we had issues with software breakpoints and used `gdb_breakpoint_override hard`. There were also issues with openocd versions, I think it required >= 0.12

let me double-check both. I should also check if openocd is part of the Zephyr SDK. If not, I should add a hint about the missing dependency and required version.

dnltz commented:
> I haven't tried this yet, but I remember we had issues with software breakpoints and used `gdb_breakpoint_override hard`. There were also issues with openocd versions, I think it required >= 0.12

works fine for me. Did you try something else @gramsay0? 

```
(gdb) break main
Breakpoint 1 at 0x140: file /home/daniel/work/zephyr/zephyr/samples/hello_world/src/main.c, line 11.
(gdb) c
Continuing.

Breakpoint 1, main () at /home/daniel/work/zephyr/zephyr/samples/hello_world/src/main.c:11
11		printf("Hello World! %s\n", CONFIG_BOARD_TARGET);
(gdb) s
13		return 0;
(gdb) s
bg_thread_main (unused1=<optimized out>, unused2=<optimized out>, unused3=<optimized out>) at /home/daniel/work/zephyr/zephyr/kernel/init.c:462
462		z_thread_essential_clear(&z_main_thread);
(gdb) s
z_thread_essential_clear (thread=0x2478 <z_main_thread>)
    at /home/daniel/work/zephyr/zephyr/kernel/include/kthread.h:176
176		thread->base.user_options &= ~K_ESSENTIAL;
(gdb) 
```

dnltz committed:
boards: phytec: phyboard_lyra: Add OpenOCD support for M4

The phyBOARD-Lyra AM62x is equipped with an XDS110 JTAG debugger. Add
missing OpenOCD configuration and some documentation to support
'west debug'.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

gramsay0 commented:
> works fine for me. Did you try something else @gramsay0?

It's been a while since I have tried it... Those quirks may have been related to A53 cores. It seems to be working well the way you have it set up. Thanks for testing!

aleeraser commented:
I have a TI SK-AM62 board and I wanted to test it, but I see a weird behavior.

When compiling whichever project e.g. the hello world, using as build target `-b sk_am62/am6234/m4`, the linking phase (i.e. the very last step) hangs forever.

```sh
...
[2/151] Generating include/generated/version.h
-- Zephyr version: 3.6.99 (/home/alessandro/zephyrproject/zephyr), build: v3.6.0-2963-g11384877d74a
[150/151] Linking C executable zephyr/zephyr.elf ^C
ninja: build stopped: interrupted by user.
```

I analyzed this further and I think the offending line is `board_set_flasher_ifnset(openocd)`.
The same also happens if I comment out the line added by this PR, since it is redundant with the content of `${ZEPHYR_BASE}/boards/common/openocd.board.cmake`.

dnltz commented:
> > I have a TI SK-AM62 board and I wanted to test this PR, but I see a weird behavior.
> 
> I just let it run for a while and I realized it completes after ~2 minutes, while a normal build takes ~6.5s. Is this expected?

you're right. I know adding configs like `CONFIG_OUTPUT_DISASSEMBLY` increases the linker time quite a lot, which seems to be related to gcc's linker. Let me check if openocd pulls in some configs/requirements.

dnltz commented:
> > I have a TI SK-AM62 board and I wanted to test this PR, but I see a weird behavior.
> 
> I just let it run for a while and I realized it completes after ~2 minutes, while a normal build takes ~6.5s. Is this expected?

It took 2 minutes to fill gaps in the binary. `CONFIG_BUILD_NO_GAP_FILL` disables this steps. It works without gaps on the phyBOARD-Lyra. I would assume the same result on the EVM.

gramsay0 commented on boards/phytec/phyboard_lyra/phyboard_lyra_am6234_m4_defconfig:
```diff
@@ -5,6 +5,9 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+# Don't fill gaps to increase linking time
```
Consider rewording this, it decreases the post build hex file generation time.
But maybe something about non-contiguous sections producing unwieldy binary and gap filled hex files would be clearer

dnltz replied:
Thanks! Is the new wording better?

gramsay0 replied:
Not really, I wasn't very clear...

The increased build time is due to a post build step generating a 7GB hex file.

This is due to the output image having non-contiguous sections (~2.5GB apart):
```
  [ 1] rom_start         PROGBITS        00000000 0000d8 000140 00 WAX  0   0  4
  ...
  [36] .resource_table   PROGBITS        9cc00000 00decc 000010 00  WA  0   0  4
```
Text/rodata/data/bss in sram0 (0x0).
openamp resource table in ddr0 (0x9cc00000).

Maybe something like this?

> The output image has non-contiguous sections, disable
gap-fill to avoid very large hex files and long build times

The output binary is already disabled for this same reason too:
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/ti/k3/am6x/Kconfig.defconfig#L42

dnltz replied:
this explanation was very helpful, thanks. I just enabled the CONFIG for all AM6 M4s since we cannot flash a hex anyways (at least right now).

dnltz committed:
soc: ti: k3: Enable BUILD_NO_GAP_FILL for all M4

By default, a post build step tries to fill gaps in the output
hex file. Since the AM62x, for example, has different,
non-contiguous memory sections, it tries to fill a gap over
multiple GBs. Disable this feature since the K3 architecture has no
dedicated flash for the firmware and stores it in a Linux rootfs.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented:
Hey @mbolivar-ampere - can you please take a look?
